File: server/src/server/statistics.py
from typing import ClassVar

import litestar
from litestar import Controller
from msgspec import Struct

from db.repositories.scores import ScoreRepo


class WaifuScoreResult(Struct):
    bucket: str
    count: int


class StatisticsController(Controller):
    path = "/statistics"
    tags: ClassVar[list[str]] = ["Statistics"]  # type: ignore

    @litestar.get("/")
    async def get_waifu_scorer_statistics(self, scores: ScoreRepo) -> list[WaifuScoreResult]:
        """Histogram of waifu scores in 10 integer-wide buckets ``[0,1), …, [9,10]``."""
        distribution = await scores.waifu_score_distribution()
        return [
            WaifuScoreResult(bucket=f"{b}~{b + 1}", count=c)
            for b, c in distribution
        ]

File: apps/api/src/routes/statistics.ts
/** `/v2/statistics` —— waifu 分数直方图。 */
import { createRoute, OpenAPIHono, z } from '@hono/zod-openapi'
import { waifuScoreDistribution } from '@pictoria/db'
import { getDb } from '../db.js'
import { OK, zodErrorHook } from '../openapi.js'

const WaifuScoreResult = z
  .object({ bucket: z.string(), count: z.int() })
  .openapi('WaifuScoreResult')

export const statisticsRoutes = new OpenAPIHono({ defaultHook: zodErrorHook })

statisticsRoutes.openapi(
  createRoute({
    method: 'get',
    path: '/v2/statistics',
    operationId: 'v2GetWaifuScorerStatistics',
    summary: 'GetWaifuScorerStatistics',
    description: 'Histogram of waifu scores in 10 integer-wide buckets ``[0,1), …, [9,10]``.',
    responses: {
      200: { description: OK, content: { 'application/json': { schema: z.array(WaifuScoreResult) } } },
    },
  }),
  (c) => {
    const { sqlite } = getDb()
    // 标签形如 "0~1" … "9~10"。
    const body = waifuScoreDistribution(sqlite).map(({ bucket, count }) => ({
      bucket: `${bucket}~${bucket + 1}`,
      count,
    }))
    return c.json(body)
  },
)
